File: class148_BacktrackingAndPalindromes/Code06_SuperPalindromesII.py
"""
超级回文数II问题 - Python版本

问题描述：
如果一个正整数自身是回文数，而且它也是一个回文数的平方，那么我们称这个数为超级回文数。
给定两个正整数 L 和 R（以字符串形式表示），返回包含在范围 [L, R] 中的超级回文数的最小值和最大值。
如果不存在超级回文数，返回[-1, -1]。

算法思路：
方法1：枚举法 - 通过生成回文数来优化搜索
方法2：打表法 - 预计算所有可能的超级回文数

时间复杂度分析：
- 枚举法：O(√R * log R)，其中√R为平方根范围，log R为回文判断时间
- 打表法：预处理O(K * log K)，查询O(log K)，其中K为超级回文数个数（约70）

空间复杂度分析：
- 枚举法：O(1)，常数额外空间
- 打表法：O(K)，存储预计算的超级回文数

工程化考量：
1. 大数处理：使用Python内置大整数支持
2. 边界处理：处理L和R的边界情况
3. 性能优化：选择合适的算法策略
4. 可测试性：设计全面的测试用例
"""

from typing import List, Tuple

class SuperPalindromesIISolver:
    def __init__(self):
        # 预计算的超级回文数数组
        self.super_palindromes = [
            1, 4, 9, 121, 484, 10201, 12321, 14641, 40804, 44944,
            1002001, 1234321, 4008004, 100020001, 102030201, 104060401,
            121242121, 123454321, 125686521, 400080004, 404090404,
            10000200001, 10221412201, 12102420121, 12345654321, 40000800004,
            1000002000001, 1002003002001, 1004006004001, 1020304030201,
            1022325232201, 1024348434201, 1210024200121, 1212225222121,
            1214428244121, 1232346432321, 1234567654321, 4000008000004,
            4004009004004, 100000020000001, 100220141022001, 102012040210201,
            102234363432201, 121000242000121, 121242363242121, 123212464212321,
            123456787654321, 400000080000004, 10000000200000001, 10002000300020001,
            10004000600040001, 10020210401202001, 10022212521222001, 10024214841242001,
            10201020402010201, 10203040504030201, 10205060806050201, 10221432623412201,
            10223454745432201, 12100002420000121, 12102202520220121, 12104402820440121,
            12122232623222121, 12124434743442121, 12321024642012321, 12323244744232321,
            12343456865434321, 12345678987654321, 40000000800000004, 40004000900040004
        ]
    
    def superpalindromes_in_range(self, left: str, right: str) -> List[int]:
        """
        方法1：打表法
        使用预计算的超级回文数列表进行查询
        """
        L = int(left)
        R = int(right)
        min_val = float('inf')
        max_val = float('-inf')
        found = False
        
        # 查找范围内的最小值和最大值
        for num in self.super_palindromes:
            if L <= num <= R:
                found = True
                if num < min_val:
                    min_val = num
                if num > max_val:
                    max_val = num
        
        if not found:
            return [-1, -1]
        
        return [int(min_val), int(max_val)]
    
    def superpalindromes_in_range_generate(self, left: str, right: str) -> List[int]:
        """
        方法2：枚举生成法
        通过生成回文数种子来构造超级回文数
        """
        L = int(left)
        R = int(right)
        min_val = float('inf')
        max_val = float('-inf')
        found = False
        
        seed_limit = 100000  # 10^5
        
        # 枚举所有可能的种子
        for seed in range(1, seed_limit):
            # 生成偶数长度的回文数
            even_pal = self.even_enlarge(seed)
            square_even = even_pal * even_pal
            
            # 检查是否在范围内且是回文数
            if L <= square_even <= R and self.is_palindrome(square_even):
                found = True
                if square_even < min_val:
                    min_val = square_even
                if square_even > max_val:
                    max_val = square_even
            
            # 生成奇数长度的回文数
            odd_pal = self.odd_enlarge(seed)
            square_odd = odd_pal * odd_pal
            
            # 检查是否在范围内且是回文数
            if L <= square_odd <= R and self.is_palindrome(square_odd):
                found = True
                if square_odd < min_val:
                    min_val = square_odd
                if square_odd > max_val:
                    max_val = square_odd
            
            # 如果生成的平方已经超过R，可以提前终止
            if square_even > R and square_odd > R:
                break
        
        if not found:
            return [-1, -1]
        
        return [int(min_val), int(max_val)]
    
    def even_enlarge(self, seed: int) -> int:
        """
        根据种子扩充到偶数长度的回文数字
        例如：seed=123，返回123321
        """
        ans = seed
        temp = seed
        while temp > 0:
            ans = ans * 10 + temp % 10
            temp //= 10
        return ans
    
    def odd_enlarge(self, seed: int) -> int:
        """
        根据种子扩充到奇数长度的回文数字
        例如：seed=123，返回12321
        """
        ans = seed
        temp = seed // 10
        while temp > 0:
            ans = ans * 10 + temp % 10
            temp //= 10
        return ans
    
    def is_palindrome(self, x: int) -> bool:
        """
        判断一个数是否是回文数
        使用数学方法避免字符串转换
        """
        if x < 0:
            return False
        if x < 10:
            return True
        
        original = x
        reversed_num = 0
        
        while x > 0:
            reversed_num = reversed_num * 10 + x % 10
            x //= 10
        
        return original == reversed_num

# 补充训练题目 - Python实现

def largest_palindrome_product(n: int) -> int:
    """
    LeetCode 479. 最大回文数乘积
    给定一个整数n，返回可表示为两个n位整数乘积的最大回文整数
    """
    if n == 1:
        return 9
    
    max_num = 10**n - 1
    min_num = 10**(n - 1)
    
    for i in range(max_num, min_num - 1, -1):
        # 创建回文数
        s = str(i)
        palindrome = int(s + s[::-1])
        
        # 检查是否可以分解为两个n位数的乘积
        for j in range(max_num, int(palindrome**0.5) - 1, -1):
            if palindrome % j == 0:
                factor = palindrome // j
                if min_num <= factor <= max_num:
                    return palindrome
    
    return -1

def count_super_palindromes(left: str, right: str) -> int:
    """
    LeetCode 906. 超级回文数（统计版本）
    统计范围内的超级回文数个数
    """
    L = int(left)
    R = int(right)
    count = 0
    
    super_palindromes = [
        1, 4, 9, 121, 484, 10201, 12321, 14641, 40804, 44944,
        1002001, 1234321, 4008004, 100020001, 102030201, 104060401,
        121242121, 123454321, 125686521, 400080004, 404090404,
        10000200001, 10221412201, 12102420121, 12345654321, 40000800004
    ]
    
    for num in super_palindromes:
        if L <= num <= R:
            count += 1
    
    return count

def is_palindrome_big(s: str) -> bool:
    """
    LeetCode 9. 回文数（大数版本）
    支持大数判断的回文数函数
    """
    left, right = 0, len(s) - 1
    while left < right:
        if s[left] != s[right]:
            return False
        left += 1
        right -= 1
    return True

def generate_palindromes_in_range(start: int, end: int) -> List[int]:
    """
    生成指定范围内的所有回文数
    """
    result = []
    
    for i in range(start, end + 1):
        if str(i) == str(i)[::-1]:
            result.append(i)
    
    return result

def next_palindrome(n: int) -> int:
    """
    寻找大于n的最小回文数
    """
    def is_palindrome(x: int) -> bool:
        s = str(x)
        return s == s[::-1]
    
    while True:
        n += 1
        if is_palindrome(n):
            return n

def prev_palindrome(n: int) -> int:
    """
    寻找小于n的最大回文数
    """
    def is_palindrome(x: int) -> bool:
        s = str(x)
        return s == s[::-1]
    
    while n > 0:
        n -= 1
        if is_palindrome(n):
            return n
    return 0

def palindrome_generator(limit: int):
    """
    回文数生成器
    生成小于等于limit的所有回文数
    """
    for i in range(1, limit + 1):
        if str(i) == str(i)[::-1]:
            yield i

# 测试函数
def test_super_palindromes_ii():
    """测试函数"""
    solver = SuperPalindromesIISolver()
    
    # 测试用例1
    left1 = "4"
    right1 = "1000"
    result1 = solver.superpalindromes_in_range(left1, right1)
    result1_gen = solver.superpalindromes_in_range_generate(left1, right1)
    
    print("测试用例1:")
    print(f"范围: [{left1}, {right1}]")
    print(f"打表法结果: [{result1[0]}, {result1[1]}]")
    print(f"生成法结果: [{result1_gen[0]}, {result1_gen[1]}]")
    print()
    
    # 测试用例2：无超级回文数的情况
    left2 = "1000000000000000000"
    right2 = "1000000000000000000"
    result2 = solver.superpalindromes_in_range(left2, right2)
    
    print("测试用例2:")
    print(f"范围: [{left2}, {right2}]")
    print(f"结果: [{result2[0]}, {result2[1]}]")
    print()
    
    # 测试补充题目
    print("=== 补充训练题目测试 ===")
    
    # 测试最大回文数乘积
    print(f"最大回文数乘积(n=2): {largest_palindrome_product(2)}")
    
    # 测试超级回文数统计
    print(f"超级回文数统计: 范围[4, 1000] -> {count_super_palindromes('4', '1000')}")
    
    # 测试大数回文判断
    big_num = "12345678987654321"
    print(f"大数回文判断: {big_num} -> {'是' if is_palindrome_big(big_num) else '否'}")
    
    # 测试回文数生成
    palindromes = generate_palindromes_in_range(100, 200)
    print(f"回文数生成(100-200): 数量={len(palindromes)}")
    
    # 测试下一个回文数
    n = 123
    print(f"下一个回文数: {n} -> {next_palindrome(n)}")
    
    # 测试上一个回文数
    print(f"上一个回文数: {n} -> {prev_palindrome(n)}")
    
    # 测试回文数生成器
    print("回文数生成器(1-100):", end=" ")
    count = 0
    for p in palindrome_generator(100):
        count += 1
    print(f"数量={count}")

if __name__ == "__main__":
    test_super_palindromes_ii()

"""
算法技巧总结 - Python版本

核心概念：
1. 回文数生成技术：
   - 种子生成法：通过种子数字构造回文数
   - 对称性利用：利用回文数的对称特征
   - 数学方法：避免字符串转换的开销

2. Python特有优势：
   - 内置大整数支持：无需担心数值溢出
   - 字符串操作简便：[::-1]快速反转字符串
   - 生成器表达式：节省内存空间

3. 算法选择策略：
   - 小范围查询：枚举法更节省内存
   - 多次查询：打表法性能最优
   - 大数据范围：需要数学优化和边界处理

调试技巧：
1. 使用pdb进行调试
2. 打印中间状态变量
3. 使用assert进行条件验证

性能优化：
1. 避免不必要的字符串转换
2. 使用局部变量减少属性查找
3. 利用Python内置函数的高效实现

工程化实践：
1. 类型注解：提高代码可读性
2. 异常处理：确保程序健壮性
3. 单元测试：保证代码质量
4. 文档字符串：提供清晰的接口说明

边界情况处理：
1. 空范围：返回[-1, -1]
2. 单个数字：检查该数字是否是超级回文数
3. 大数范围：使用适当的算法避免性能问题
4. 边界值：测试最小/最大可能的值
"""

File: class148_BacktrackingAndPalindromes/Code06_SuperPalindromesII.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <climits>
#include <limits>

using namespace std;

/**
 * 超级回文数II问题 - C++版本
 * 
 * 问题描述：
 * 如果一个正整数自身是回文数，而且它也是一个回文数的平方，那么我们称这个数为超级回文数。
 * 给定两个正整数 L 和 R（以字符串形式表示），返回包含在范围 [L, R] 中的超级回文数的最小值和最大值。
 * 如果不存在超级回文数，返回[-1, -1]。
 * 
 * 算法思路：
 * 方法1：枚举法 - 通过生成回文数来优化搜索
 * 方法2：打表法 - 预计算所有可能的超级回文数
 * 
 * 时间复杂度分析：
 * - 枚举法：O(√R * log R)，其中√R为平方根范围，log R为回文判断时间
 * - 打表法：预处理O(K * log K)，查询O(log K)，其中K为超级回文数个数（约70）
 * 
 * 空间复杂度分析：
 * - 枚举法：O(1)，常数额外空间
 * - 打表法：O(K)，存储预计算的超级回文数
 * 
 * 工程化考量：
 * 1. 大数处理：使用long long类型避免溢出
 * 2. 边界处理：处理L和R的边界情况
 * 3. 性能优化：选择合适的算法策略
 * 4. 可测试性：设计全面的测试用例
 */

class SuperPalindromesIISolver {
public:
    /**
     * 方法1：枚举法
     * 通过生成回文数来优化搜索，避免直接遍历大范围
     */
    vector<long long> superpalindromesInRange(string left, string right) {
        long long L = stoll(left);
        long long R = stoll(right);
        long long minVal = numeric_limits<long long>::max();
        long long maxVal = numeric_limits<long long>::min();
        bool found = false;
        
        // 预计算的超级回文数数组
        vector<long long> superPalindromes = getSuperPalindromes();
        
        // 查找范围内的最小值和最大值
        for (long long num : superPalindromes) {
            if (num >= L && num <= R) {
                found = true;
                if (num < minVal) minVal = num;
                if (num > maxVal) maxVal = num;
            }
        }
        
        if (!found) {
            vector<long long> result;
            result.push_back(-1);
            result.push_back(-1);
            return result;
        }
        
        vector<long long> result;
        result.push_back(minVal);
        result.push_back(maxVal);
        return result;
    }
    
    /**
     * 方法2：枚举生成法
     * 通过生成回文数种子来构造超级回文数
     */
    vector<long long> superpalindromesInRangeGenerate(string left, string right) {
        long long L = stoll(left);
        long long R = stoll(right);
        long long minVal = numeric_limits<long long>::max();
        long long maxVal = numeric_limits<long long>::min();
        bool found = false;
        
        long long seedLimit = 100000; // 10^5
        
        // 枚举所有可能的种子
        for (long long seed = 1; seed < seedLimit; seed++) {
            // 生成偶数长度的回文数
            long long evenPal = evenEnlarge(seed);
            long long squareEven = evenPal * evenPal;
            
            // 检查是否在范围内且是回文数
            if (squareEven >= L && squareEven <= R && isPalindrome(squareEven)) {
                found = true;
                if (squareEven < minVal) minVal = squareEven;
                if (squareEven > maxVal) maxVal = squareEven;
            }
            
            // 生成奇数长度的回文数
            long long oddPal = oddEnlarge(seed);
            long long squareOdd = oddPal * oddPal;
            
            // 检查是否在范围内且是回文数
            if (squareOdd >= L && squareOdd <= R && isPalindrome(squareOdd)) {
                found = true;
                if (squareOdd < minVal) minVal = squareOdd;
                if (squareOdd > maxVal) maxVal = squareOdd;
            }
            
            // 如果生成的平方已经超过R，可以提前终止
            if (squareEven > R && squareOdd > R) {
                break;
            }
        }
        
        if (!found) {
            vector<long long> result;
            result.push_back(-1);
            result.push_back(-1);
            return result;
        }
        
        vector<long long> result;
        result.push_back(minVal);
        result.push_back(maxVal);
        return result;
    }
    
private:
    /**
     * 获取预计算的超级回文数列表
     */
    vector<long long> getSuperPalindromes() {
        return {
            1LL, 4LL, 9LL, 121LL, 484LL, 10201LL, 12321LL, 14641LL, 40804LL, 44944LL,
            1002001LL, 1234321LL, 4008004LL, 100020001LL, 102030201LL, 104060401LL,
            121242121LL, 123454321LL, 125686521LL, 400080004LL, 404090404LL,
            10000200001LL, 10221412201LL, 12102420121LL, 12345654321LL, 40000800004LL,
            1000002000001LL, 1002003002001LL, 1004006004001LL, 1020304030201LL,
            1022325232201LL, 1024348434201LL, 1210024200121LL, 1212225222121LL,
            1214428244121LL, 1232346432321LL, 1234567654321LL, 4000008000004LL,
            4004009004004LL, 100000020000001LL, 100220141022001LL, 102012040210201LL,
            102234363432201LL, 121000242000121LL, 121242363242121LL, 123212464212321LL,
            123456787654321LL, 400000080000004LL, 10000000200000001LL, 10002000300020001LL,
            10004000600040001LL, 10020210401202001LL, 10022212521222001LL, 10024214841242001LL,
            10201020402010201LL, 10203040504030201LL, 10205060806050201LL, 10221432623412201LL,
            10223454745432201LL, 12100002420000121LL, 12102202520220121LL, 12104402820440121LL,
            12122232623222121LL, 12124434743442121LL, 12321024642012321LL, 12323244744232321LL,
            12343456865434321LL, 12345678987654321LL, 40000000800000004LL, 40004000900040004LL
        };
    }
    
    /**
     * 根据种子扩充到偶数长度的回文数字
     */
    long long evenEnlarge(long long seed) {
        long long ans = seed;
        long long temp = seed;
        while (temp > 0) {
            ans = ans * 10 + temp % 10;
            temp /= 10;
        }
        return ans;
    }
    
    /**
     * 根据种子扩充到奇数长度的回文数字
     */
    long long oddEnlarge(long long seed) {
        long long ans = seed;
        long long temp = seed / 10;
        while (temp > 0) {
            ans = ans * 10 + temp % 10;
            temp /= 10;
        }
        return ans;
    }
    
    /**
     * 判断一个数是否是回文数
     */
    bool isPalindrome(long long x) {
        if (x < 0) return false;
        if (x < 10) return true;
        
        long long original = x;
        long long reversed = 0;
        
        while (x > 0) {
            reversed = reversed * 10 + x % 10;
            x /= 10;
        }
        
        return original == reversed;
    }
};

/**
 * 补充训练题目 - C++实现
 */

/**
 * 判断一个数是否是回文数（全局函数版本）
 */
bool isPalindrome(long long x) {
    if (x < 0) return false;
    if (x < 10) return true;
    
    long long original = x;
    long long reversed = 0;
    
    while (x > 0) {
        reversed = reversed * 10 + x % 10;
        x /= 10;
    }
    
    return original == reversed;
}

/**
 * LeetCode 479. 最大回文数乘积
 * 给定一个整数n，返回可表示为两个n位整数乘积的最大回文整数
 */
long long largestPalindrome(int n) {
    if (n == 1) return 9;
    
    long long maxNum = 1;
    for (int i = 0; i < n; i++) {
        maxNum *= 10;
    }
    maxNum -= 1;
    
    long long minNum = 1;
    for (int i = 0; i < n - 1; i++) {
        minNum *= 10;
    }
    
    for (long long i = maxNum; i >= minNum; i--) {
        // 创建回文数
        string s = to_string(i);
        string rev = s;
        reverse(rev.begin(), rev.end());
        long long palindrome = stoll(s + rev);
        
        // 检查是否可以分解为两个n位数的乘积
        for (long long j = maxNum; j * j >= palindrome; j--) {
            if (palindrome % j == 0) {
                long long factor = palindrome / j;
                if (factor >= minNum && factor <= maxNum) {
                    return palindrome;
                }
            }
        }
    }
    
    return -1;
}

/**
 * LeetCode 906. 超级回文数（统计版本）
 * 统计范围内的超级回文数个数
 */
int countSuperPalindromes(string left, string right) {
    long long L = stoll(left);
    long long R = stoll(right);
    int count = 0;
    
    vector<long long> superPalindromes = {
        1LL, 4LL, 9LL, 121LL, 484LL, 10201LL, 12321LL, 14641LL, 40804LL, 44944LL,
        1002001LL, 1234321LL, 4008004LL, 100020001LL, 102030201LL, 104060401LL,
        121242121LL, 123454321LL, 125686521LL, 400080004LL, 404090404LL,
        10000200001LL, 10221412201LL, 12102420121LL, 12345654321LL, 40000800004LL
    };
    
    for (long long num : superPalindromes) {
        if (num >= L && num <= R) {
            count++;
        }
    }
    
    return count;
}

/**
 * LeetCode 9. 回文数（大数版本）
 * 支持大数判断的回文数函数
 */
bool isPalindromeBig(string s) {
    int left = 0, right = s.length() - 1;
    while (left < right) {
        if (s[left] != s[right]) {
            return false;
        }
        left++;
        right--;
    }
    return true;
}

/**
 * 生成指定范围内的所有回文数
 */
vector<long long> generatePalindromesInRange(long long start, long long end) {
    vector<long long> result;
    
    for (long long i = start; i <= end; i++) {
        if (isPalindrome(i)) {
            result.push_back(i);
        }
    }
    
    return result;
}

// 测试函数
void testSuperPalindromesII() {
    SuperPalindromesIISolver solver;
    
    // 测试用例1
    string left1 = "4";
    string right1 = "1000";
    vector<long long> result1 = solver.superpalindromesInRange(left1, right1);
    vector<long long> result1_gen = solver.superpalindromesInRangeGenerate(left1, right1);
    
    cout << "测试用例1:" << endl;
    cout << "范围: [" << left1 << ", " << right1 << "]" << endl;
    cout << "打表法结果: [" << result1[0] << ", " << result1[1] << "]" << endl;
    cout << "生成法结果: [" << result1_gen[0] << ", " << result1_gen[1] << "]" << endl;
    cout << endl;
    
    // 测试用例2：无超级回文数的情况
    string left2 = "1000000000000000000";
    string right2 = "1000000000000000000";
    vector<long long> result2 = solver.superpalindromesInRange(left2, right2);
    
    cout << "测试用例2:" << endl;
    cout << "范围: [" << left2 << ", " << right2 << "]" << endl;
    cout << "结果: [" << result2[0] << ", " << result2[1] << "]" << endl;
    cout << endl;
    
    // 测试补充题目
    cout << "=== 补充训练题目测试 ===" << endl;
    
    // 测试最大回文数乘积
    cout << "最大回文数乘积(n=2): " << largestPalindrome(2) << endl;
    
    // 测试超级回文数统计
    cout << "超级回文数统计: 范围[4, 1000] -> " << countSuperPalindromes("4", "1000") << endl;
    
    // 测试大数回文判断
    string bigNum = "12345678987654321";
    cout << "大数回文判断: " << bigNum << " -> " << (isPalindromeBig(bigNum) ? "是" : "否") << endl;
    
    // 测试回文数生成
    vector<long long> palindromes = generatePalindromesInRange(100, 200);
    cout << "回文数生成(100-200): 数量=" << palindromes.size() << endl;
}

int main() {
    testSuperPalindromesII();
    return 0;
}

/**
 * 算法技巧总结 - C++版本
 * 
 * 核心概念：
 * 1. 回文数生成技术：
 *    - 种子生成法：通过种子数字构造回文数
 *    - 对称性利用：利用回文数的对称特征
 *    - 数学方法：避免字符串转换的开销
 * 
 * 2. 算法选择策略：
 *    - 小范围查询：枚举法更节省内存
 *    - 多次查询：打表法性能最优
 *    - 大数据范围：需要数学优化和边界处理
 * 
 * 3. 性能优化技巧：
 *    - 提前终止：当结果超出范围时提前结束搜索
 *    - 数学优化：使用位运算和数学方法
 *    - 缓存策略：预计算常用结果
 * 
 * 调试技巧：
 * 1. 边界值测试：最小/最大输入、边界情况
 * 2. 中间结果打印：跟踪算法执行过程
 * 3. 性能分析：使用profiler工具分析瓶颈
 * 
 * 工程化实践：
 * 1. 模块化设计：分离算法逻辑和业务逻辑
 * 2. 异常安全：确保资源正确释放
 * 3. 代码可读性：使用有意义的变量名和注释
 */